[CodeGen] SparseLiveVariables: A liveness analysis pass

hiraditya opened the pull request:
`SparseLiveVariables` computes register liveness using `SparseBitVector` and a `LivenessTracker` API. It performs a single backward walk of a basic block to determine virtual register kills, providing liveness state for the analyzed blocks with O(N) complexity per block.

Key Components:
1. LivenessTracker: Computes and caches the killed virtual registers per instruction during a backward block traversal.
2. APIs: Provides `updateLiveIns` and `updateKillFlags` to allow consumer passes to query and synchronize MIR liveness state.
3. Fallback Mechanism: Includes a `hasAnalyzed` API, allowing passes to safely fall back to conservative heuristics for blocks that were created dynamically after the analysis was run (e.g., critical edge splitting).

Integration into MachineLICM to replace its register pressure estimation. 
- `MachineLICM::calcRegisterCost` now queries the cached `LivenessTracker` for blocks that have been analyzed, avoiding inaccurate register pressure estimates.
- To preserve existing aggressive loop-invariant hoisting behavior, the integration retains the `MRI->hasOneNonDBGUse(Reg)` heuristic alongside the new liveness queries.
- In AArch64, this allows `uunpk` instruction to be hoisted (`dag-combine-concat-vectors.ll`). It also prevents constants from being incorrectly hoisted across loop backedges by calculating the register pressure cost (`pr164181.ll`).

hiraditya committed:
Add tests

hiraditya committed:
Add live variable analysis pass

hiraditya committed:
WIP genset calculation

hiraditya committed:
Verify liveness is working

hiraditya committed:
Add flag for live variable analysis

Add an mir test

hiraditya committed:
clang format

hiraditya commented:
Forgot the issue where there was a discussion on how liveness analysis will help. Please tag the ticket if someone remembers.

hiraditya committed:
Update kill of individual instructions

hiraditya committed:
Add test from #166141

hiraditya committed:
clang-format and add correct flags to the tests

hiraditya committed:
Use uses and defs iterator to make sure defs are processed before uses

hiraditya committed:
Update test cases to stop after live variables

hiraditya committed:
Update test to check loop invariant motion

hiraditya committed:
Comment

lenary commented:
How RISC-V specific is this pass? Should it maybe be called something like MachineLiveness, and put in a target-independent directory?

I don't mind if we do any of that generalisation work as a follow-up, but it would be good to know the trajectory of this.

lukel97 reviewed (commented):
Does this compute anything different from the LiveVariables analysis?

hiraditya commented:
> Do you intend to make this a full analysis pass that other passes can query?

That can be done, but it can be a bit tricky if we update the kill flags as that would change the IR. Are there  analysis passes that modify the IR?

hiraditya committed:
Return true when kill information is updated

hiraditya committed:
Add a flag to disable verification

hiraditya committed:
Update live-in set of basic blocks

hiraditya committed:
Add a flag to gate update of live-in set

hiraditya committed:
Add tests for post-regalloc liveness analysis

topperc commented on llvm/lib/Target/RISCV/RISCVLiveVariables.cpp:
```diff
@@ -0,0 +1,598 @@
+  const RISCVSubtarget &Subtarget = MF.getSubtarget<RISCVSubtarget>();
+
+  // Verify we're targeting RV64
+  if (!Subtarget.is64Bit()) {
```
We should use the triple in RISCVTargetMachine to not insert the pass for RV32.

topperc commented on llvm/lib/Target/RISCV/RISCVLiveVariables.cpp:
```diff
@@ -0,0 +1,598 @@
+  std::unordered_map<unsigned, unsigned> TrackedRegisters;
+
+  /// Liveness information for each basic block
+  DenseMap<const MachineBasicBlock *, LivenessInfo> BlockLiveness;
```
Should this be a vector using basic block numbering to index?

hiraditya replied:
Yeah we can do that as well, but wouldn't that require renumbering? I'm fine either way, lmk if you have a preference.

FWIW, the DenseMap here doesn't introduce non-determinism as there is no transformation that iterates upon this.

arsenm commented on llvm/lib/Target/RISCV/RISCVLiveVariables.cpp:
```diff
@@ -0,0 +1,598 @@
+// in RISC-V machine functions, optimized for RV64 (64-bit RISC-V architecture).
+//
+// The analysis performs a backward dataflow analysis to compute
+// liveness information. Also updates the kill flags on register operands.
```
Kill flags are deprecated, should not be introducing new code computing or depending on them. We should be deleting the current LiveVariables pass

arsenm reviewed (commented):
What is this for that LiveIntervals can't be used for?

asb commented:
@hiraditya I took a look to try to benchmark this - I'm not able to build SPEC, multiple failures of:
```
Computed live-in set is inconsistent with MBB.
UNREACHABLE executed at ../../llvm/lib/Target/RISCV/RISCVLiveVariables.cpp:443!
```

I'll attempt to get a minimised test case for you.

hiraditya commented:
> @hiraditya I took a look to try to benchmark this - I'm not able to build SPEC, multiple failures of:
> ```
> Computed live-in set is inconsistent with MBB.
> UNREACHABLE executed at ../../llvm/lib/Target/RISCV/RISCVLiveVariables.cpp:443!
> ```
> 
> I'll attempt to get a minimised test case for you.

Oh this one is because the current liveness information is stale. I wrote the verifier to prove that(see https://github.com/llvm/llvm-project/issues/168202). We can either:

 1. disable verifier (-mllvm -riscv-liveness-verify), more conservative approach as it keeps the existing basic block live in/outs
2. Enable (-mllvm -riscv-liveness-update-mbb-liveness) to update the basic block live in/out sets.

hiraditya committed:
Update default flags

hiraditya committed:
Use TargetTriple to bail out for RV32

asb commented:
What is your expectation for the impact because I've tried this with SPEC and I've not succeeded in getting any meaningful difference in dynamic instruction count for SPEC2017 `-march=rva22u64 -O3` (i.e. the diff rounds to 0.00%). All below done with the 2e0659bfe0ff506a5ece9f2b3fc1ee397bef3345 version of the PR.

Also, please remember to give guidance on what combinations of flags you expect people to use in order to test this properly! My attempts:

* `-mllvm -riscv-enable-live-variables`
  * Compiler crashes as reported above
* `-mllvm -riscv-enable-live-variables -mllvm -riscv-liveness-verify=0`
  * Compiles but no diff in generated .s vs baseline
* `-riscv-enable-live-variables -mllvm -riscv-liveness-update-mbb-liveins`
  * 8961 insertions, 8959 deletions in .s output. No impact on dyn inst count above 0.00% (rounded)
* `-mllvm -riscv-enable-live-variables -mllvm -riscv-liveness-update-mbb-liveins -mllvm -riscv-liveness-update-kills`
  * 19273 insertions, 19022 deletions in .s output. Reduction in dyn inst count of 0.01% for povray, all others round to 0.00%.

hiraditya commented on llvm/lib/Target/RISCV/RISCVLiveVariables.cpp:
```diff
@@ -0,0 +1,598 @@
+    Info.Gen.clear();
+    Info.Use.clear();
+
+    // Process instructions in forward order to build Use and Def sets
```
Can you point to a reference? AIUI, within a BasicBlock, a forward traversal would find the use-set (live-in) and build from there.

hiraditya committed:
Add test from the bug report

arsenm reviewed (commented):
Maintaining yet another liveness system is a very substantial maintenance burden.

- This should not be RISCV specific, from the beginning
- There should be a concerted effort to delete the existing LiveVariables

hiraditya committed:
Fix incorrect RegCounter increment

hiraditya committed:
Make pass target independent

hiraditya committed:
Make it an analysis pass and document

hiraditya committed:
Bug fixes triggered by test

hiraditya committed:
clang-format

arsenm commented on llvm/include/llvm/CodeGen/SparseLiveVariables.h:
```diff
@@ -0,0 +1,224 @@
+        : LiveRegs(LiveOut), MRI(MRI) {}
+
+    void stepBackward(const MachineInstr &MI) {
+      if (MI.isDebugInstr() || MI.isMetaInstruction())
```
Meta instruction liveness cannot be ignored. There are meta instructions that exist specifically to manage the liveness (e.g., kill and implicit_def)

arsenm commented on llvm/include/llvm/CodeGen/SparseLiveVariables.h:
```diff
@@ -0,0 +1,224 @@
+      }
+
+      if (!MI.isPHI()) {
+        for (const MachineOperand &MO : MI.operands()) {
```
Fuse these into one loop over operands?

arsenm commented on llvm/include/llvm/CodeGen/SparseLiveVariables.h:
```diff
@@ -0,0 +1,224 @@
+  /// Returns the computed Live-In set for the given MachineBasicBlock.
+  const SparseBitVector<> &getLiveInSet(const MachineBasicBlock *MBB) const {
+    auto It = BlockLiveness.find(MBB);
+    assert(It != BlockLiveness.end() && "Block not analyzed");
```
.at?

arsenm commented on llvm/include/llvm/CodeGen/SparseLiveVariables.h:
```diff
@@ -0,0 +1,224 @@
+  /// Returns the computed Live-Out set for the given MachineBasicBlock.
+  const SparseBitVector<> &getLiveOutSet(const MachineBasicBlock *MBB) const {
+    auto It = BlockLiveness.find(MBB);
+    assert(It != BlockLiveness.end() && "Block not analyzed");
```
.at?

arsenm commented on llvm/lib/CodeGen/MachineLICM.cpp:
```diff
@@ -768,9 +777,31 @@ bool MachineLICMImpl::IsGuaranteedToExecute(MachineBasicBlock *BB,
+
+  SparseLiveVariables::LivenessTracker Tracker(LV->getLiveOutSet(MBB), MRI);
+  for (const MachineInstr &MI : llvm::reverse(*MBB)) {
+    if (!MI.isDebugInstr() && !MI.isMetaInstruction()) {
```
Can't skip meta instructions

arsenm commented on llvm/lib/CodeGen/SparseLiveVariables.cpp:
```diff
@@ -0,0 +1,509 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
```
Needs long description here

arsenm commented on llvm/lib/CodeGen/SparseLiveVariables.cpp:
```diff
@@ -0,0 +1,509 @@
+
+      for (const MachineBasicBlock *Succ : MBB->successors())
+        if (Reachable.count(Succ))
+          BlockLiveness[MBB].LiveOut |= BlockLiveness[Succ].LiveIn;
```
Map lookup in loop

arsenm commented on llvm/lib/CodeGen/SparseLiveVariables.cpp:
```diff
@@ -0,0 +1,509 @@
+      for (MachineInstr &MI : llvm::reverse(*MBB))
+        Tracker.stepBackward(MI);
+
+      BlockLiveness[MBB].LiveIn = Tracker.getLiveSet();
```
Repeating map lookups

arsenm commented on llvm/lib/CodeGen/SparseLiveVariables.cpp:
```diff
@@ -0,0 +1,509 @@
+  }
+}
+
+void SparseLiveVariables::updateKillFlags(MachineFunction &MF) const {
```
Kill flags have been deprecated-ish for > 10 years. Shouldn't put extra effort into maintaining them

arsenm reviewed (commented):
What's the plan here? Are you going to use this to complete the long-overdue deletion of LiveVariables?

hiraditya committed:
Address PR comments on SparseLiveVariables

- Do not ignore meta instructions
- Fuse loop over operands
- Use DenseMap::at instead of find

hiraditya committed:
Remove deprecated updateKillFlags method

As per PR review feedback, kill flags are deprecated and we shouldn't
be computing or depending on them. This removes the unused
updateKillFlags method and its associated test.

hiraditya committed:
Address remaining PR comments for SparseLiveVariables

- Replaced DenseMap with std::vector for BlockLiveness, indexing by MBB->getNumber()
  to resolve undefined behavior and improve performance.
- Removed redundant reachability checks.
- Documented vector gap behavior and overall pass design.

hiraditya commented:
> What's the plan here? Are you going to use this to complete the long-overdue deletion of LiveVariables?

i can if there is still interest in removing the current LiveVariable. Currently i'm working towards leveraging this in MachineLICM and likely other backend optimizations.

arsenm commented:
> i can if there is still interest in removing the current LiveVariable. Currently i'm working towards leveraging this in MachineLICM and likely other backend optimizations.

I think It's quite important to remove LiveVariables

hiraditya committed:
Add full unit test coverage and fix LiveOut PHI bug

hiraditya committed:
Support new Pass Manager

hiraditya commented:
> I think It's quite important to remove LiveVariables

Since that'd be a large change, should i put that as part of separate PR? There are probably 5 stages to that

1. The core LLVM codegen passes (PHIElimination, TwoAddressInstructionPass etc) currently manually update and query LiveVariables. Transition them to use SparseLiveVariables
2. Remove Legacy LiveVariables
3. Remove Kill Flags from Target Backends
4. If passes suffer compile-time regressions due to calling SparseLiveVariables::isKillAt() repeatedly, we will fix those specific passes to query liveness more optimally rather than introducing a generic cache.
5. Put out a RFC for removing MachineOperand::isKill bit

arsenm commented:
> 1. The core LLVM codegen passes (PHIElimination, TwoAddressInstructionPass etc) currently manually update and query LiveVariables. Transition them to use SparseLiveVariables

I believe TwoAddressInstructionPass is the one true user of LiveVariables, and only for tied instructions. I'm not sure what PHIElimination is doing with it. I'd probably start by transitioning these to the replacement

hiraditya commented:
> I believe TwoAddressInstructionPass is the one true user of LiveVariables, and only for tied instructions. I'm not sure what PHIElimination is doing with it. I'd probably start by transitioning these to the replacement

makes sense, I'm looking into it.

This PR is ready to be reviewed btw.

hiraditya committed:
[SparseLiveVariables] Enable analysis for MachineLICM

Enable the SparseLiveVariables analysis pass for MachineLICM.
This properly fetches the analysis under both the legacy pass manager
and the new pass manager.

hiraditya commented:
Follow up PR to remove LiveVariables: https://github.com/hiraditya/llvm-project/pull/1 it depends on this PR.
